src/FSMCIO.cpp

```diff
@@ -0,0 +1,172 @@
+#include "FSMCIO.h"
+#include "dma.h"
+
+#define ZERO(f) memset(&f, 0, sizeof(f))
+#define oops() target_panic(909)
+
+#ifdef FSMC_Bank1
+
+namespace codal
+{
+
+static void FMC_BANK1_Init(uint32_t flags, SRAM_HandleTypeDef &hsram)
+{
+    FMC_NORSRAM_TimingTypeDef sram_timing;
+    FMC_NORSRAM_TimingTypeDef sram_timing_write;
+
+    /*** Configure the SRAM Bank 1 ***/
+    /* Configure IPs */
+    hsram.Instance = FSMC_Bank1;
+    hsram.Extended = FSMC_Bank1E;
+
+    /* Timing for READING */
+    sram_timing.AddressSetupTime = 9;
+    sram_timing.AddressHoldTime = 1;
+    sram_timing.DataSetupTime = 36;
+    sram_timing.BusTurnAroundDuration = 1;
+    sram_timing.CLKDivision = 2;
+    sram_timing.DataLatency = 2;
+    sram_timing.AccessMode = FSMC_ACCESS_MODE_A;
+
+    /* Timing for WRITTING*/
+    sram_timing_write.AddressSetupTime = 0;
+    sram_timing_write.AddressHoldTime = 1;
+    // 4 gives 72ns write cycle, datasheet for ILI9341 says 66ns min
+    // 3 gives 60ns
+    // everything works down to 1
+    sram_timing_write.DataSetupTime = flags & 0x1f;
+    sram_timing_write.BusTurnAroundDuration = 0;
+    sram_timing_write.CLKDivision = 2;
+    sram_timing_write.DataLatency = 2;
+    sram_timing_write.AccessMode = FSMC_ACCESS_MODE_A;
+
+    hsram.Init.NSBank = FSMC_NORSRAM_BANK1;
+    hsram.Init.DataAddressMux = FSMC_DATA_ADDRESS_MUX_DISABLE;
+    hsram.Init.MemoryType = FSMC_MEMORY_TYPE_SRAM;
+    hsram.Init.MemoryDataWidth = FSMC_NORSRAM_MEM_BUS_WIDTH_8;
+    hsram.Init.BurstAccessMode = FSMC_BURST_ACCESS_MODE_DISABLE;
+    hsram.Init.WaitSignalPolarity = FSMC_WAIT_SIGNAL_POLARITY_LOW;
+    hsram.Init.WrapMode = FSMC_WRAP_MODE_DISABLE;
+    hsram.Init.WaitSignalActive = FSMC_WAIT_TIMING_BEFORE_WS;
+    hsram.Init.WriteOperation = FSMC_WRITE_OPERATION_ENABLE;
+    hsram.Init.WaitSignal = FSMC_WAIT_SIGNAL_DISABLE;
+    hsram.Init.ExtendedMode = FSMC_EXTENDED_MODE_ENABLE;
+    hsram.Init.AsynchronousWait = FSMC_ASYNCHRONOUS_WAIT_DISABLE;
+    hsram.Init.WriteBurst = FSMC_WRITE_BURST_DISABLE;
+    hsram.Init.WriteFifo = FSMC_WRITE_FIFO_DISABLE;
+    hsram.Init.PageSize = FSMC_PAGE_SIZE_NONE;
+    hsram.Init.ContinuousClock = FSMC_CONTINUOUS_CLOCK_SYNC_ONLY;
+
+    HAL_SRAM_Init(&hsram, &sram_timing, &sram_timing_write);
+}
+
+static FSMCIO *theInstance;
+
+FSMCIO::FSMCIO(uint32_t flags, PinNumber wr, PinNumber rd)
+{
+    ZERO(hsram);
+    ZERO(hdma);
+
+    theInstance = this;
+
+    GPIO_InitTypeDef gpio_init_structure;
```

Comment:
This is all pretty ugly low level stuff. I guess it doesn't really make sense to use the pin abstraction for all these GPIO because it costs some RAM for no reason.